### DISCUSSION TITLE
How to repair a pdf?

### DISCUSSION THREAD

**Discussion by workflowsguy** [pikepdf/pikepdf#511] [Q&A]
The documentation briefly mentions that with `pikepdf`, it is also possible to repair a pdf file.

I see that there is a `pdf.check()` function,  however, I could not find any explanation how to proceed when this function returns warnings or errors.

Thank you,

Guy

**Comment by jbarlow83** [comment]
pikepdf auto-repairs many files. It's hard to say what is wrong with your file without more detail, and not all files are repairable.

When the error is that some referenced object is missing, for example, there is probably no way to repair that, unless you can regenerate the PDF from its source documents.

**Comment by workflowsguy** [comment]
Thanks for the clarification and the explanation.

I am currently using the `qpdf` command line tools to filter out potentially defective PDFs with `qpdf --check` and repair them like so:

```
qpdf -qdf "$PDFFILE" "$QDFFILE"
fix-qdf "$QDFFILE" > "$QDFFILENEW"
qpdf "$QDFFILENEW" "$PDFFILENEW"
```

I was hoping to replace the script with a pure Python solution, however:

- `pikepdf`'s `Pdf.check()` function might not report any issues on a file whereas `qpdf --check` does
- after opening and then saving a file with issues reported by `Pdf.check()`, the resulting file still had the same issues. Only after opening and saving the PDF with PyMuPDF were no issues reported with the new file.

**Comment by jbarlow83** [comment]
Since `Pdf.check()` is nearly the same code as `qpdf --check` (both using libqpdf) that is surprising.

I can't debug an issue without a reproducing file.

**Comment by workflowsguy** [comment]
Over the last days, I ran a verification script on several hundred PDF files and found 10 with this issue.

Unfortunately, they all contain copyrighted material so I can not share them. What I can provide is a list of the type of warnings `qpdf --check` produces for those files where `Pdf.check` did not report anything.

Maybe they can provide you some insight until I encounter a PDF file that I *can* share...

**Types of warnings from `qpdf --check` (occuring from 1 to n times per file)**

- first page object offset mismatch
- first shared object number mismatch: hint table = 2313; computed = 2317
- incorrect object count in outline hint table
- incorrect offset in outlines table: hint table = 115397; computed = 73267
- linearized file contains an uncompressed object after a compressed one in a cross-reference stream
- modify annotations: not allowed
- modify anything: not allowed
- modify document assembly: not allowed
- modify forms: not allowed
- modify other: not allowed
- object count mismatch for page 0: hint table = 1490; computed = 1633
- page 0 has shared identifier entries
- page 0: shared object 14075: in hint table but not computed list
- page 1: shared object 20498: in computed list but not hint table
- page length mismatch for page 0: hint table = 109162; computed length = 108120 (offset = 3214)
- print high resolution: allowed
- print low resolution: allowed
- shared object 0 length mismatch: hint table = 133; computed = 224
- Supplied password is user password
- User password = 


Tested with pikepdf 8.4.0 and qpdf 11.5.0

Should I create an issue for this?

**Comment by jbarlow83** [comment]
I think it's that pikepdf.Pdf.open() defaults to enabling repairs, so it fixes many of the issues qpdf --check would otherwise report and fix.

Those errors are all fairly innocuous, except "page 0 has shared identifier entries".

**Comment by workflowsguy** [comment]
Converting this into an issue after I finally found some sample pdf's that I believe can be shared.

**Comment by philipspeter696-hub** [comment]
Repair corrupt, broken, or highly damaged PDF files effortlessly with Recoveryfix PDF Repair Software. This advanced tool restores inaccessible PDF documents while preserving original formatting, text, images, and structure. It efficiently fixes corruption caused by system crashes, malware, or file damage. Enjoy a fast, secure, and reliable PDF recovery solution with an easy-to-use interface for all users.
